File: samples/FsBulletML2.Sample.MonoGame.FSharp/Enemy.fs
namespace FsBulletML2.Sample.MonoGame.FSharp

open System.Collections.Generic
open Microsoft.Xna.Framework
open FsBulletML2
open FsBulletML2.MonoGame

type Enemy(life) as this =
    inherit BaseBullet()

    [<DefaultValue>]
    val mutable private self: IEnemy

    [<DefaultValue>]
    val mutable private timer: int

    [<DefaultValue>]
    val mutable private bulletName: string

    [<DefaultValue>]
    val mutable private bulletBulletmlInfo: BulletmlInfo

    [<DefaultValue>]
    val mutable private second: bool

    [<DefaultValue>]
    val mutable private bullet: EnemyBullet

    [<DefaultValue>]
    val mutable private life: int32

    new() = Enemy(2000)

    do
        this.self <- this :> IEnemy
        this.self.BulletType <- BulletType.Enemy
        this.self.Init()
        this.self.IsBullet <- false
        this.self.Radius <- 18.f
        this.life <- life

    member this.Timer = this.timer

    interface IEnemy with
        member this.Life
            with get () = this.life
            and set (v) = this.life <- v

        member this.Shoot() =
            let self = this :> IEnemy

            if self.Used then
                this.bullet <- new EnemyBullet()
                (this.bullet :> IBullet).IsBullet <- true
                Manager.addEnemyBulletPos (this.bullet, Vector2(self.X, self.Y))
                this.bullet.SetScript(Some(this.bulletBulletmlInfo.Script(loadRand, loadRank ())))

        member this.Update() =
            this.timer <- this.timer + 1

            let finish =
                if this.bullet :> obj = null then
                    false
                else
                    (this.bullet :> IBullet).Finished

            if not this.second || finish then
                this.second <- true
                this.timer <- 0
                let self = this :> IEnemy
                self.Shoot()

            let apply x y =
                this.self.X <- this.self.X + x
                this.self.Y <- this.self.Y + y

            base.RunTask(System.Action<_, _>(apply))

    member this.SetMoveBulletmlInfo(bulletmlInfo: BulletmlInfo) =
        (this :> IEnemy).SetScript(Some(bulletmlInfo.Script(loadRand, loadRank ())))

    member this.SetBulletTask(bulletName, bulletmlInfo) =
        this.bulletName <- bulletName
        this.bulletBulletmlInfo <- bulletmlInfo

module EnemyControl =
    let bullets = FsBulletML2.Bullets.Dsl.All.bullets

File: samples/FsBulletML2.Sample.MonoGame.CSharp/Enemy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Microsoft.Xna.Framework;

using FsBulletML2;
using FsBulletML2.MonoGame;
using BulletType = FsBulletML2.DTD.BulletType;
using EnemyBullet = FsBulletML2.MonoGame.EnemyBullet;

namespace FsBulletML2.Sample.MonoGame.CSharp
{
    public class Enemy : FsBulletML2.MonoGame.BaseBullet, IEnemy 
    {
        private FsBulletML2.MonoGame.IBullet self = null;
        private int Timer { get; set; }
        string BulletName { get; set; }
        BulletmlInfo BulletmlInfo { get; set; }
        private bool Second { get; set; }
        EnemyBullet Bullet { get; set; }
        public int Life { get; set; }

        public Enemy() : this(2000) {}
        public Enemy(int life)
        {
            this.self = this;
            this.self.BulletType = BulletType.Enemy;
            self.Init();
            this.self.IsBullet = false;
            this.self.Radius = 18;
            this.Life = life;
        }

        public void Shoot()
        {
          if (this.self.Used)
          {
            this.Bullet = new EnemyBullet();
            ((IBullet)this.Bullet).IsBullet = true;
            Manager.AddEnemyBulletPos(this.Bullet, new Vector2(self.X, self.Y));
            this.Bullet.SetScript(this.BulletmlInfo.Script(BulletmlLoad.loadEnv()));
          }
        }

        void IBullet.Update() { this.Update(); }

        public void Update()
        { 
            this.Timer += 1;

            if (!this.Second || IsFinish()) 
            {
                this.Second = true;
                this.Timer = 0;
                this.Shoot();
            }
            base.RunTask((x, y) => { this.self.X = this.self.X + x; this.self.Y = this.self.Y + y; });
        }

        private bool IsFinish()
        {
            if (this.Bullet == null)
            {
                return false;
            }
            else
            {
                return ((IBullet)this.Bullet).Finished;
            }
       
        }

        public void SetMoveBulletmlInfo(BulletmlInfo bulletmlInfo) 
        {
            ((IEnemy)this).SetScript(bulletmlInfo.Script(BulletmlLoad.loadEnv()), null);
        }

        public void SetBulletmlInfo(string bulletName, BulletmlInfo bulletmlInfo) 
        {
            this.BulletName = bulletName;
            this.BulletmlInfo = bulletmlInfo;
        }

    }
}
